op-service/dial/active_l2_provider_test.go

```diff
@@ -0,0 +1,744 @@
+	require.NoError(t, err)
+	require.Same(t, primarySequencer, firstSequencerUsed)
+
+	active, err := primarySequencer.SequencerActive(context.Background())
+	require.NoError(t, err)
+	require.False(t, active)
+
+	secondSequencerUsed, err := rollupProvider.RollupClient(context.Background())
+	require.NoError(t, err)
+	require.Same(t, primarySequencer, secondSequencerUsed)
+	ept.assertAllExpectations(t)
+}
+
+// TestEndpointProvider_ReturnsSameSequencerOnInactiveWithLongCheckDuration verifies that the ActiveL2EndpointProvider
+// still returns the same sequencer across calls even if it becomes inactive, due to a long check duration.
+func TestEndpointProvider_ReturnsSameSequencerOnInactiveWithLongCheckDuration(t *testing.T) {
+	ept := setupEndpointProviderTest(t, 2)
+	primarySequencer := ept.rollupClients[0]
+
+	longCheckDuration := 1 * time.Hour
+	primarySequencer.ExpectSequencerActive(true, nil) // Active on creation
+
+	endpointProvider, err := ept.newActiveL2EndpointProvider(longCheckDuration)
+	require.NoError(t, err)
+
+	// Primary sequencer becomes inactive, but the provider won't check immediately due to longCheckDuration
+	primarySequencer.ExpectSequencerActive(false, nil)
+	firstEthClientUsed, err := endpointProvider.EthClient(context.Background())
+	require.NoError(t, err)
+	require.Same(t, ept.ethClients[0], firstEthClientUsed)
+
+	active, err := primarySequencer.SequencerActive(context.Background())
+	require.NoError(t, err)
+	require.False(t, active)
+
+	secondEthClientUsed, err := endpointProvider.EthClient(context.Background())
+	require.NoError(t, err)
+	require.Same(t, ept.ethClients[0], secondEthClientUsed)
+	ept.assertAllExpectations(t)
+}
+
+// TestRollupProvider_HandlesManyIndexClientMismatch verifies that the ActiveL2RollupProvider avoids
+// the case where the index of the current sequencer does not match the index of the current rollup client.
+func TestRollupProvider_HandlesManyIndexClientMismatch(t *testing.T) {
+	ept := setupEndpointProviderTest(t, 3)
+	seq0, seq1, seq2 := ept.rollupClients[0], ept.rollupClients[1], ept.rollupClients[2]
+
+	// "start happy": primarySequencer is active on creation
+	seq0.ExpectSequencerActive(true, nil) // active on creation
+	rollupProvider, err := ept.newActiveL2RollupProvider(0)
+	require.NoError(t, err)
+
+	// primarySequencer goes down
+	seq0.ExpectSequencerActive(false, fmt.Errorf("I'm offline now"))
+	seq0.MaybeClose()
+	ept.setRollupDialOutcome(0, false) // primarySequencer fails to dial
+	// secondarySequencer is inactive, but online
+	seq1.ExpectSequencerActive(false, nil)
+	seq1.MaybeClose()
+	// tertiarySequencer can't even be dialed
+	ept.setRollupDialOutcome(2, false)
+	// after calling RollupClient, index will be set to 0, but currentRollupClient is secondarySequencer
+	rollupClient, err := rollupProvider.RollupClient(context.Background())
+	require.Error(t, err)
+	require.Nil(t, rollupClient)
+	// internal state would now be inconsistent in the buggy impl.
```

Comment:
```suggestion
	// internal state would now be inconsistent in a buggy impl.
```
since it's now more general about any would-be buggy impl